src/websocket_tracker_connection.cpp

```diff
@@ -0,0 +1,399 @@
+	if(req.event != event_t::none)
+		payload["event"] = event_string[static_cast<int>(req.event) - 1];
+
+	payload["peer_id"] = from_latin1({req.pid.data(), req.pid.size()});
+
+	payload["offers"] = json::array();
+	for(auto const& offer : req.offers)
+	{
+		json payload_offer;
+		payload_offer["offer_id"] = from_latin1({offer.id.data(), offer.id.size()});
+		payload_offer["offer"]["type"] = "offer";
+		payload_offer["offer"]["sdp"] = offer.sdp;
+		payload["offers"].push_back(payload_offer);
+	}
+
+	std::string const data = payload.dump();
+
+	std::shared_ptr<request_callback> cb = requester();
+	if (cb) cb->debug_log("*** WEBSOCKET_TRACKER_WRITE [ %s ]", data.c_str());
+
+	std::shared_ptr<websocket_tracker_connection> self(shared_from_this());
+    m_websocket->async_write_some(boost::asio::const_buffer(data.data(), data.size())
+     		, std::bind(&websocket_tracker_connection::on_write, self, _1, _2));
+}
+
+void websocket_tracker_connection::do_send(tracker_answer const& ans)
+{
+    json payload;
+    payload["action"] = "announce";
+    payload["info_hash"] = from_latin1({ans.info_hash.data(), std::size_t(ans.info_hash.size())});
+    payload["offer_id"] = from_latin1({ans.answer.offer_id.data(), ans.answer.offer_id.size()});
+    payload["to_peer_id"] = from_latin1({ans.answer.pid.data(), ans.answer.pid.size()});
+    payload["peer_id"] =  from_latin1({ans.pid.data(), ans.pid.size()});
+    payload["answer"]["type"] = "answer";
+    payload["answer"]["sdp"] = ans.answer.sdp;
+
+	std::string const data = payload.dump();
+
+	std::shared_ptr<request_callback> cb = requester();
+	if (cb) cb->debug_log("*** WEBSOCKET_TRACKER_WRITE [ %s ]", data.c_str());
+
+	std::shared_ptr<websocket_tracker_connection> self(shared_from_this());
+	m_websocket->async_write_some(boost::asio::const_buffer(data.data(), data.size())
+			, std::bind(&websocket_tracker_connection::on_write, self, _1, _2));
+}
+
+void websocket_tracker_connection::do_read()
+{
+	m_read_buffer.clear();
+
+	std::shared_ptr<websocket_tracker_connection> self(shared_from_this());
+	m_websocket->async_read_some(m_read_buffer
+            , std::bind(&websocket_tracker_connection::on_read, self, _1, _2));
+}
+
+void websocket_tracker_connection::on_connect(error_code const& ec)
+{
+	if(ec)
+	{
+		std::shared_ptr<request_callback> cb = requester();
+		if (cb) cb->debug_log("*** WEBSOCKET_TRACKER_FAILED [ url: %s, error: %d ]"
+				, tracker_req().url.c_str(), int(ec.value()));
+
+		while(!m_pending.empty()) m_pending.pop();
+		return;
+	}
+
+	std::shared_ptr<request_callback> cb = requester();
+    if (cb) cb->debug_log("*** WEBSOCKET_TRACKER_CONNECTED [ url: %s ]"
+    		, tracker_req().url.c_str());
+
+	send_pending();
+	do_read();
+}
+
+void websocket_tracker_connection::on_timeout(error_code const& /*ec*/)
+{
+	// Dummy
+}
+
+void websocket_tracker_connection::on_read(error_code const& ec, std::size_t /* bytes_read */)
+{
+	if(ec)
+    {
+        // Ignore
+        return;
```

Comment:
why are errors being ignored? shouldn't this report back an error to the callback issuing the announce?

Reply:
Indeed, I'm calling `tracker_request_error()` instead.

Reply:
this will terminate the connection entirely. Is there logic that attempts to re-establish this? Technically, since all WebTorrent announces go through this connection (iiuc), any queued up tracker announce should fail here too (I think). But I'm not entirely sure where those are queued up.